openverse_catalog/dags/providers/provider_csv_load_scripts/inaturalist/taxa.sql

```diff
@@ -3,51 +3,165 @@
+      and added it manually, so use that.
+    - Otherwise, use the iNaturalist name. There are a few where the name is "Not
+      assigned" but we're going to filter those records out & drop associated photos.
+
+Tags:
+    - If the title of a specific taxa is a vernacular name, put the iNaturalist name in
+      the tags.
+    - If there are non-English vernacular names, put them in the tags.
+    - Put the titles of ancestors in the tags.
+    - Given the order of types of tags above plus alphabetical order, take only the
+      first 20 tags.
+
+Representing tags in this way to be consistent with python processing `_enrich_tags`:
+TO DO #902: Find a DRYer way to do this enrichment with SQL
 */
 
-DROP TABLE IF EXISTS inaturalist.taxa;
-COMMIT;
+/*
+                   ********** Create tag type ***********
+This at least makes the structure of tags a little more explicit in sql
+*/
+DO $$ BEGIN
+    create type openverse_tag as (name varchar(255), provider varchar(255));
+EXCEPTION
+    WHEN duplicate_object THEN null;
+END $$;
 
+/*              ********** Load raw iNaturalist Data *********                */
+DROP TABLE IF EXISTS inaturalist.taxa;
 CREATE TABLE inaturalist.taxa (
     taxon_id integer,
     ancestry character varying(255),
     rank_level double precision,
     rank character varying(255),
     name character varying(255),
-    active boolean,
-    ancestor_names varchar[]
+    active boolean
 );
-COMMIT;
-
 SELECT aws_s3.table_import_from_s3('inaturalist.taxa',
     'taxon_id, ancestry, rank_level, rank, name, active',
     '(FORMAT ''csv'', DELIMITER E''\t'', HEADER, QUOTE E''\b'')',
     'inaturalist-open-data',
     'taxa.csv.gz',
     'us-east-1');
-
 ALTER TABLE inaturalist.taxa ADD PRIMARY KEY (taxon_id);
 COMMIT;
 
-WITH aggregated AS
+/*
+                ********** Integrate Catalog of Life Data *********
+
++ aggregate the catalog of life vernacular names table to the scientific name level,
+  with string lists of English names and JSON for tags in other languages
++ add in information from the manual table of common names and from inaturalist taxa,
+  so that every taxon record has the best possible title
++ add in inaturalist taxa base data, with the category of "Life" excluded from
+  ancestry strings
+*/
+drop table if exists inaturalist.taxa_with_vernacular;
+create table inaturalist.taxa_with_vernacular as
+(
+    with
+    catalog_of_life_names as
+    (
+        select cast(md5(n.scientificname) as uuid) as md5_scientificname,
+            v.name_language,
+            v.taxon_name,
+            row_number() OVER (partition by n.scientificname
+                order by case when v.name_language='eng' then 0 else 1 end,
+                v.taxon_name)
+                as name_rank,
+            count(case when v.name_language='eng' then 1 end)
+                OVER (partition by n.scientificname
+                rows between unbounded preceding and unbounded following)
+                as english_names
```

Comment:
Do you mind adding comments to each of these two lines describing what they're doing? I'm having trouble parsing it out 😅

Reply:
I added a bunch of comments, but one key question which might make this all simpler: I got the impression somewhere that the python code I was trying to mimic was taking only the first 20 tags for any image. Did I make that up somehow? Should I just take all the tags? (Just FYI, with vernacular names, a few taxa would have thousands of tags.)

Reply:
Ah, I don't think we have a limit on tags currently! There is this note in Flickr though:

https://github.com/WordPress/openverse-catalog/blob/5ae98b0190c8d0971ac900b053ce1bd60dcaacc8/openverse_catalog/dags/providers/provider_api_scripts/flickr.py#L251-L253

I think 20 is good, but we should definitely note it somewhere!

Reply:
+1 for limiting to 20!

Reply:
Ah, ok and so then the question becomes, is it practical or just absurdly xenophobic to write code to prioritize / select for English language vernacular in those first 20? I thought about just doing some kind of alphabetical order or leaving order out of it entirely (though I need to check if `row_number()` requires an `order by`)... ???

Reply:
That's....a good question 😅 Do you know if they already appear in some hierarchical order or anything?